discovery/sync_manager.go

```diff
@@ -0,0 +1,637 @@
+// candidate syncers.
+//
+// NOTE: This method requires the syncersMtx read lock to be held.
+func (m *SyncManager) chooseRandomSyncer(
+	blacklist map[routing.Vertex]struct{}) *GossipSyncer {
+
+	for _, s := range m.syncers {
+		if blacklist != nil {
+			if _, ok := blacklist[s.cfg.peerPub]; ok {
+				continue
+			}
+		}
+		if s.syncState() != chansSynced {
+			continue
+		}
+		return s
+	}
+	return nil
+}
+
+// InitSyncState is called by outside sub-systems when a connection is
+// established to a new peer that understands how to perform channel range
+// queries. We'll allocate a new gossip syncer for it, and start any goroutines
+// needed to handle new queries.
+//
+// TODO(wilmer): Only mark as ActiveSync if this isn't a channel peer.
+func (m *SyncManager) InitSyncState(peer lnpeer.Peer) {
+	m.syncersMtx.Lock()
+	defer m.syncersMtx.Unlock()
+
+	// If we already have a syncer, then we'll exit early as we don't want
+	// to override it.
+	nodeID := routing.Vertex(peer.PubKey())
+	if _, ok := m.gossipSyncer(nodeID); ok {
+		return
+	}
+
+	log.Infof("Creating new GossipSyncer for peer=%x", nodeID[:])
+
+	encoding := lnwire.EncodingSortedPlain
+	s := newGossipSyncer(GossipSyncerCfg{
+		chainHash:     m.cfg.ChainHash,
+		peerPub:       nodeID,
+		channelSeries: m.cfg.ChanSeries,
+		encodingType:  encoding,
+		chunkSize:     encodingTypeToChunkSize[encoding],
+		sendToPeer: func(msgs ...lnwire.Message) error {
+			return peer.SendMessage(false, msgs...)
+		},
+	})
+
+	// If we don't have any active syncers, we'll default to forcing a full
+	// sync with the first peer we connect to, to ensure we get as much of
+	// the graph as possible. This is mostly done for light clients, which
+	// shouldn't be receiving new updates anyway due to the high bandwidth
+	// consumption that stems from it.
+	if m.cfg.NumActiveSyncers == 0 && len(m.syncers) == 0 {
+		s.syncType = uint32(FullSync)
+		m.syncers[nodeID] = s
+		s.Start()
+		return
+	}
+
+	// If we've yet to reach our desired number of active syncers, then
+	// we'll use this one.
+	if len(m.activeSyncers) < m.cfg.NumActiveSyncers {
+		s.syncType = uint32(ActiveSync)
+		m.activeSyncers[nodeID] = s
+		m.enqueueActiveSyncer(s, false)
+		return
+	}
+
+	// Otherwise, the syncer should be initialized as passive.
+	s.syncType = uint32(PassiveSync)
+	s.state = uint32(waitingQueryRangeReply)
+	m.syncers[nodeID] = s
+	s.Start()
+}
+
+// PruneSyncState is called by outside sub-systems once a peer that we were
+// previously connected to has been disconnected. In this case we can stop the
+// existing GossipSyncer assigned to the peer and free up resources.
+func (m *SyncManager) PruneSyncState(peer routing.Vertex) {
+	m.syncersMtx.Lock()
+	defer m.syncersMtx.Unlock()
+
+	s, ok := m.gossipSyncer(peer)
+	if !ok {
+		return
+	}
+
+	log.Infof("Removing GossipSyncer for peer=%v", peer)
+
+	// We'll start by stopping the GossipSyncer for the disconnected peer.
+	s.Stop()
```

Comment:
should we free the lock while waiting for the syncer to stop?

Reply:
Don't think it'll be much of a problem, but agree that it is nice. Fixed.